wayland-server/src/dispatch.rs

```diff
@@ -274,27 +274,30 @@ impl<I: Resource + 'static, U: Send + Sync + 'static, D: Dispatch<I, UserData =
 /// ```
 #[macro_export]
 macro_rules! delegate_dispatch {
-    ($dispatch_from: ty: [$($interface: ty),*] => $dispatch_to: ty) => {
-        $(
-            impl $crate::Dispatch<$interface> for $dispatch_from {
-                type UserData = <$dispatch_to as $crate::DelegateDispatchBase<$interface>>::UserData;
+    (@impl $dispatch_from:ident $(< $( $lt:tt $( : $clt:tt $(+ $dlt:tt )* )? ),+ >)? : $interface:ty => $dispatch_to: ty) => {
+        impl$(< $( $lt $( : $clt $(+ $dlt )* )? ),+ >)? $crate::Dispatch<$interface> for $dispatch_from$(< $( $lt ),+ >)? {
+            type UserData = <$dispatch_to as $crate::DelegateDispatchBase<$interface>>::UserData;
 
-                fn request(
-                    &mut self,
-                    client: &$crate::Client,
-                    resource: &$interface,
-                    request: <$interface as $crate::Resource>::Request,
-                    data: &Self::UserData,
-                    dhandle: &mut $crate::DisplayHandle<'_>,
-                    data_init: &mut $crate::DataInit<'_, Self>,
-                ) {
-                    <$dispatch_to as $crate::DelegateDispatch<$interface, Self>>::request(self, client, resource, request, data, dhandle, data_init)
-                }
+            fn request(
+                &mut self,
+                client: &$crate::Client,
+                resource: &$interface,
+                request: <$interface as $crate::Resource>::Request,
+                data: &Self::UserData,
+                dhandle: &mut $crate::DisplayHandle<'_>,
+                data_init: &mut $crate::DataInit<'_, Self>,
+            ) {
+                <$dispatch_to as $crate::DelegateDispatch<$interface, Self>>::request(self, client, resource, request, data, dhandle, data_init)
+            }
 
-                fn destroyed(&mut self, client: $crate::backend::ClientId, resource: $crate::backend::ObjectId, data: &Self::UserData) {
-                    <$dispatch_to as $crate::DelegateDispatch<$interface, Self>>::destroyed(self, client, resource, data)
-                }
+            fn destroyed(&mut self, client: $crate::backend::ClientId, resource: $crate::backend::ObjectId, data: &Self::UserData) {
+                <$dispatch_to as $crate::DelegateDispatch<$interface, Self>>::destroyed(self, client, resource, data)
             }
+        }
+    };
+    ($impl:tt : [$($interface: ty),*] => $dispatch_to: ty) => {
+        $(
+            $crate::delegate_dispatch!(@impl $impl : $interface => $dispatch_to);
         )*
     };
```

Comment:
Why did you need to do this split here, but not for `delegate_global_dispatch` ?

Reply:
Oh, that is a good question, maybe we need it in `delegate_global_dispatch` as well and no tests catched it?

For `delegate_dispatch` the compiler was unhappy about unrolling the repetition of `interface` without also unrolling `lt`. I didn't find another way to fix that, than splitting the macro. It was the first doc-test of `delegate_display`, that found this problem (although it appears that it is not even using multiple globals). See here: https://github.com/Smithay/wayland-rs/runs/6475799083?check_suite_focus=true#step:8:289

I have no clue why other tests are working, given that this seems to be quite the fundamental problem, but maybe `delegate_global_dispatch` is lacking a matching test? (I am not that familiar with macros, nor do I have a good understanding of wayland-rs test infrastructure. I just fiddled around with the macro definition until it seemed to work.)

Reply:
I think it is that `delegate_global_dispatch` does not have any test yeah, so I believe it'd need the same trick as `delegate_dispatch`.